oak_enclave_rt/src/heap.rs

```diff
@@ -0,0 +1,135 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+use core::{
+    alloc::{GlobalAlloc, Layout},
+    ffi::c_void,
+    ops::Deref,
+    ptr::NonNull,
+};
+use linked_list_allocator::Heap;
+use oak_restricted_kernel_interface::syscalls::{MmapFlags, MmapProtection};
+use spinning_top::Spinlock;
+
+pub struct GrowableHeap {
+    /// Underlying heap implementation.
+    heap: Heap,
+
+    /// Address of the base of the heap.
+    base: Option<usize>,
+
+    /// Address of the next virtual memory page to use for the heap.
+    cursor: Option<usize>,
+}
+
+impl GrowableHeap {
+    const PAGE_SIZE: usize = 0x20_0000usize;
+
+    pub const fn empty() -> Self {
+        Self {
+            heap: Heap::empty(),
+            base: None,
+            cursor: None,
+        }
+    }
+
+    pub unsafe fn init(&mut self) {
+        // Get the first 2 MiB of memory for the heap.
+        self.extend().unwrap();
+
+        self.heap
+            .init(self.base.unwrap() as *mut u8, Self::PAGE_SIZE);
+    }
+
+    /// Extends the current pool of memory by one 2 MiB page.
+    fn extend(&mut self) -> Result<(), &'static str> {
+        // If this is the first allocation, we don't want MAP_FIXED; however, for every consecutive
+        // allocation, we want it to come immediately after the initial allocation so we need
+        // MAP_FIXED.
+        let flags = MmapFlags::MAP_ANONYMOUS
+            | MmapFlags::MAP_PRIVATE
+            | self
+                .cursor
+                .map_or_else(MmapFlags::empty, |_| MmapFlags::MAP_FIXED);
+
+        let mem = oak_restricted_kernel_api::syscall::mmap(
+            self.cursor.map(|x| x as *const c_void),
+            Self::PAGE_SIZE as isize,
+            MmapProtection::PROT_READ | MmapProtection::PROT_WRITE,
+            flags,
+            -1,
+            0,
+        )
+        .map_err(|_| "failed to acquire memory")?;
+
+        // Move the cursor to the next unallcoated page.
+        self.cursor = Some(mem.as_ptr() as usize + Self::PAGE_SIZE);
+        self.base.get_or_insert(mem.as_ptr() as usize);
+
+        Ok(())
+    }
+
+    pub fn allocate_first_fit(&mut self, layout: Layout) -> Result<NonNull<u8>, ()> {
+        // Try allocating the data structure; if the allocation fails, grow the heap and try again
+        // until we succeed (or until we can't extend ourselves any further)
+        loop {
+            match self.heap.allocate_first_fit(layout) {
```

Comment:
Perhaps, instead of calling `allocate_first_fit` on the base allocator after every time 2MiB is added perhaps add 2MiB multiple times until more space has been added than requested in the layout and then try to allocate again. This could potentially waste a lot of space if there is just less than the required amount available at the end of the allocator, but I think that is a better tradeoff than the possible cost for this when allocating a large amount of memory with an inefficient base allocator.

Reply:
Fair enough. I've changed it to use powers of 2 when increasing the heap size.